### ISSUE TITLE
Store edge BBox in the graph

### ISSUE THREAD

**Issue by otbutz** [graphhopper/graphhopper#2625]
We currently compute the BBox on-the-fly for custom area checks and only based on the tower nodes of the edge:

https://github.com/graphhopper/graphhopper/blob/af660cdf2fc1a365b279eb6047fc5956e585cfb9/core/src/main/java/com/graphhopper/routing/weighting/custom/CustomWeightingHelper.java#L78-L85

This has a few drawbacks. On the one hand, we lose precision, since the full point list could in fact have a larger BBox than that of only the tower nodes:

![image](https://user-images.githubusercontent.com/22315436/178253713-af1f9466-ec63-4ad5-b425-4d38fe4d11c4.png)

And it adds computational overhead as discussed previously:

https://github.com/graphhopper/graphhopper/pull/2338#issuecomment-937549263
https://discuss.graphhopper.com/t/why-using-areas-is-low-efficient/7374/6

**Comment by easbar** [comment]
> We currently compute the BBox on-the-fly for custom area checks and only based on the tower nodes of the edge:

Ok, but the code where we use only the tower nodes to compute the bounding box is this method:

https://github.com/graphhopper/graphhopper/blob/06b261a3650c9f9ece9b394c65bc7dfe71a3ae04/core/src/main/java/com/graphhopper/util/GHUtility.java#L939-L944

We currently use it for the block_area feature (GraphEdgeIdFinder) and, like you say, for the custom weighting areas. We use the tower node bounding box in `if (!p.getBounds().intersects(bbox)) return false;`, which is only a performance optimization that skips the more costly detailed intersection check for many edges. We then use the full geometry for the detailed check `p.intersects(edge.fetchWayGeometry(FetchMode.ALL).makeImmutable());`.

>  On the one hand, we lose precision, since the full point list could in fact have a larger BBox than that of only the tower nodes

I agree the current logic is not very intuitive. For example if p is a rectangle and the tower node bounding box intersects p it can happen that a road is affected by a custom area even though none of the road points lies within the area (false positive). On the other hand, if the roads geometry partly lies outside of the tower node bounding box, like in your picture, it can happen that a road is **not** affected even though some points **are** located within the area (false negative).

The false positives are only possible because of the rectangle optimization introduced here: https://github.com/graphhopper/graphhopper/pull/2326, right? Without the `if (p.isRectangle()) true` case we would only ever return true if `p.intersects(edge.fetchWayGeometry(FetchMode.ALL).makeImmutable());` was true, which is only true if one of the geometry points actually lies within `p`. And I think this is what we want: A road should be affected by a custom area if and only if at least one point is located within the area?!

So to actually find out which roads are affected by a custom model area we need to do the full check `p.intersects(edge.fetchWayGeometry(FetchMode.ALL).makeImmutable());`, but yes we can skip it if the real (geometry) bounding box does not intersect the polygon envelope bounding box. Currently we already skip it if the tower node bounding box does not intersect the polygon envelope bounding box, which is wrong in some cases (the false negatives I mentioned above).

> And it adds computational overhead as discussed previously:

Actually I don't really think so. Let's say we used the full geometry bounding box in `GHUtility.createBBox` instead of the tower node bounding box we use currently. This would mean that **more** edges would be considered for the full `p.intersects(edge.fetchWayGeometry(FetchMode.ALL).makeImmutable());` check. So overall the area check could potentially be **slower**. But it would be more precise as we would get rid of the false negatives. I do not expect that reading the stored full bounding box would be any faster (or slower) than reading the tower node bounding box as we do now, because the `fetchWayGeometry(FetchMode.TOWER_ONLY)` call in `GHUtility#createBBox` does not 'fetch' the actual geometry anyway:

https://github.com/graphhopper/graphhopper/blob/06b261a3650c9f9ece9b394c65bc7dfe71a3ae04/core/src/main/java/com/graphhopper/storage/BaseGraph.java#L398-L405

I think that maybe we should make the custom area check more precise (get rid of false positives and negatives), and this would require using the real geometry bounding box instead of the tower node bounding box, but this will make the custom area feature potentially slower rather than faster. 

The code you quoted does include a (costly) `fetchWayGeometry(FetchMode.ALL)` call, but we cannot replace it with a stored bounding box, even if it was the full geometry bounding box. For this detailed check we actually want to check if any point lies within the polygon, so the bounding box wouldn't be sufficient.

So to me it seems like the real question is: Do we want to make the custom areas more precise? If yes we need to use the full bounding box for `GHUtility#createBBox` and to make sure this does not slow down the custom area feature it is probably a good idea to store the full bounding box. But overall this won't make the feature faster. If anything it will be slower.

If our goal is rather to make the custom area feature faster as requested in the forum discussion you mentioned storing the full bounding box won't help at all I think (unless we want to increase the precision without slowing it down). To speed up custom area routing we probably need to come up with something else and first we should have a reproducible example that shows how slow the routing really is.

Regarding actually storing the bounding box for each edge I found this related comment: https://github.com/graphhopper/graphhopper/pull/2320#discussion_r632417842

**Comment by easbar** [comment]
> Because especially for larger polygons the precision down to the edge geometry might not matter.

Yes it really depends on the size of the areas. For small areas the current precision seems insufficient to me, but for big ones it probably does not really matter and we could even be less precise still, i.e. only check if one of the two tower nodes is in the area or not. By the way this could be an easy optimization: We could briefly check if one of the two tower nodes is in the area before we fetch the geometry. If yes return true and only otherwise do the full `p.intersects(edge.fetchWayGeometry(FetchMode.ALL).makeImmutable());` check.

**Comment by otbutz** [comment]
> For small areas the current precision seems insufficient to me

Agreed. Blocking individual edges is a feature that is still missing in CustomModels, but it can be substituted with small areas.

Maybe we should also take a step back and get some numbers on what's really causing the slowdown here. Since people seem to be complaining, this should be measurable.
